libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -277,22 +405,91 @@ class state_history_log {
       if (block_num <= _begin_block) {
          num_removed = _end_block - _begin_block;
          log.seek(0);
-         index.seek(0);
          boost::filesystem::resize_file(log_filename, 0);
          boost::filesystem::resize_file(index_filename, 0);
          _begin_block = _end_block = 0;
       } else {
          num_removed  = _end_block - block_num;
          uint64_t pos = get_pos(block_num);
          log.seek(0);
-         index.seek(0);
          boost::filesystem::resize_file(log_filename, pos);
-         boost::filesystem::resize_file(index_filename, (block_num - _begin_block) * sizeof(uint64_t));
+         boost::filesystem::resize_file(index_filename, (block_num - _index_begin_block) * sizeof(uint64_t));
          _end_block = block_num;
       }
+      log.seek_end(0);
+      ilog("fork or replay: removed ${n} blocks from ${name}.log", ("n", num_removed)("name", name));
+   }
+
+   void vacuum() {
+      //a completely empty log should have nothing on disk; don't touch anything
+      if(_begin_block == _end_block)
+         return;
+
+      size_t copy_from_pos = get_pos(_begin_block) + sizeof(uint64_t); //skip first copied block's magic
+      //go ahead and clear the pruned flag from the log file. If we fail to completely vacuum well maybe recovery can sort out something
+      log.seek(0);
+      fc::raw::pack(log, ship_magic(ship_current_version));
+      size_t copy_to_pos = log.tellp();
+
+      //may happen if _begin_block is still first block on-disk of log. just erase 4 byte trailer
+      if(_begin_block == _index_begin_block) {
+         log.flush();
+         fc::resize_file(log.get_file_path(), fc::file_size(log.get_file_path()) - sizeof(uint32_t));
+         return;
+      }
+
+      ilog("Vacuuming pruned log ${n}", ("n", name));
+
+      const size_t offset_bytes = copy_from_pos - copy_to_pos;
+      const size_t offset_blocks = _begin_block - _index_begin_block;
+      log.seek_end(0);
+      size_t copy_sz = log.tellp() - copy_from_pos - sizeof(uint32_t); //don't copy trailer in to new unpruned log
+      const uint32_t num_blocks_in_log = _end_block - _begin_block;
+
+      std::vector<char> buff;
+      buff.resize(4*1024*1024);
+
+      auto tick = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now());
+      while(copy_sz) {
+         const size_t copy_this_round = std::min(buff.size(), copy_sz);
+         log.seek(copy_from_pos);
+         log.read(buff.data(), copy_this_round);
+         log.punch_hole(copy_from_pos, copy_from_pos+copy_this_round);
```

Comment:
Since `copy_from_pos` is changing in each iteration of the loop, doesn't this mean that the hole punching may not be contiguous if the positions are not precisely aligned with the file block size?

I understand this only matters in terms of temporary disk usage during the vacuum processing since the file is resized at the end anyway. But the temporary disk usage is why you are bothering to do this hole punching here in the first place. And the on-disk data can grow with each iteration of this loop to some fraction of the original disk size unless the positions are precisely aligned with the file block size.

What if the start of the hole punching range was instead `std::max(copy_to_pos + copy_this_round, orig_copy_from_pos)` where `orig_copy_from_pos` is the original `copy_from_pos` value before the start of the while loop? Would this be inefficient for the OS if the range is large but spans sub-ranges that have already been punched out?

Reply:
hmm.. I need to think it through a little more. Maybe even something as simple as
```c++
log.punch_hole(copy_to_pos, copy_from_pos+copy_this_round);
```
is fine.